File: HermesNotesCore/Sources/HermesNotesCore/Support/Slug.swift
import Foundation

public enum Slug {
    /// Filesystem-safe slug for markdown file names: lowercased, alphanumerics
    /// kept, everything else collapsed to single hyphens, capped in length so
    /// paths stay readable.
    public static func make(_ text: String, maxLength: Int = 60) -> String {
        var out = ""
        var lastWasHyphen = true // suppress leading hyphen
        for scalar in text.lowercased().unicodeScalars {
            if ("a"..."z").contains(String(scalar)) || ("0"..."9").contains(String(scalar)) {
                out.append(Character(scalar))
                lastWasHyphen = false
            } else if !lastWasHyphen {
                out.append("-")
                lastWasHyphen = true
            }
            if out.count >= maxLength { break }
        }
        while out.hasSuffix("-") { out.removeLast() }
        return out.isEmpty ? "untitled" : out
    }

    /// Short stable suffix from a UUID so slugs never collide.
    public static func shortID(_ id: UUID) -> String {
        String(id.uuidString.replacingOccurrences(of: "-", with: "").prefix(8)).lowercased()
    }
}

File: src/core/slug.ts
/**
 * Filesystem-safe slug for markdown file names: lowercased, alphanumerics
 * kept, everything else collapsed to single hyphens, capped in length so
 * paths stay readable.
 */
export function makeSlug(text: string, maxLength = 60): string {
  let out = '';
  let lastWasHyphen = true; // suppress leading hyphen
  for (const ch of text.toLowerCase()) {
    if (/[a-z0-9]/.test(ch)) {
      out += ch;
      lastWasHyphen = false;
    } else if (!lastWasHyphen) {
      out += '-';
      lastWasHyphen = true;
    }
    if (out.length >= maxLength) break;
  }
  out = out.replace(/-+$/, '');
  return out.length > 0 ? out : 'untitled';
}

/** Short stable suffix from a UUID so slugs never collide. */
export function shortId(id: string): string {
  return id.replace(/-/g, '').slice(0, 8).toLowerCase();
}
